File: browser/url_util.cc
// Copyright (c) 2013 mogoweb. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "c4a/browser/url_util.h"

#include "base/android/jni_string.h"
#include "base/string_util.h"
#include "jni/URLUtil_jni.h"
#include "net/base/net_util.h"
#include "url/gurl.h"

using base::android::ConvertJavaStringToUTF8;
using base::android::ConvertJavaStringToUTF16;

jstring GetSuggestedFilename(JNIEnv* env, jclass javaClass,
            jstring jurl,
            jstring jcontentDisposition,
            jstring jmimeType) {
  GURL url(ConvertJavaStringToUTF16(env, jurl));
  std::string content_disposition = ConvertJavaStringToUTF8(env, jcontentDisposition);
  std::string mime_type = ConvertJavaStringToUTF8(env, jmimeType);

  return base::android::ConvertUTF16ToJavaString(
      env,
      net::GetSuggestedFilename(url, content_disposition, EmptyString(), EmptyString(), mime_type, "download")).Release();
}

bool RegisterURLUtil(JNIEnv* env) {
  return RegisterNativesImpl(env);
}

File: browser/java/src/com/mogoweb/browser/utils/URLUtil.java
// Copyright (c) 2013 mogoweb. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package com.mogoweb.browser.utils;

public class URLUtil {

    /**
     * Guesses canonical filename that a download would have, using
     * the URL and contentDisposition. File extension, if not defined,
     * is added based on the mimetype
     * @param url Url to the content
     * @param contentDisposition Content-Disposition HTTP header or null
     * @param mimeType Mime-type of the content or null
     *
     * @return suggested filename
     */
    public static final String guessFileName(
            String url,
            String contentDisposition,
            String mimeType) {
        return nativeGetSuggestedFilename(url, contentDisposition, mimeType);
    }

    private static native String nativeGetSuggestedFilename(
            String url,
            String contentDisposition,
            String mimeType);
}
